### PR TITLE
Don't retreive pants options in Event Dispatch Thread

### LINKED ISSUE #463: Synchronous execution on EDT
```
java.lang.Throwable: Synchronous execution on EDT: /Users/yic/workspace/pants/pants options --no-colors
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:145)
	at com.intellij.execution.process.OSProcessHandler.checkEdtAndReadAction(OSProcessHandler.java:147)
	at com.intellij.execution.process.OSProcessHandler.waitFor(OSProcessHandler.java:85)
	at com.intellij.execution.process.CapturingProcessRunner.runProcess(CapturingProcessRunner.java:29)
	at com.intellij.execution.process.CapturingProcessHandler.runProcess(CapturingProcessHandler.java:55)
	at com.twitter.intellij.pants.util.PantsUtil.getCmdOutput(PantsUtil.java:705)
	at com.twitter.intellij.pants.model.PantsOptions.getPantsOptions(PantsOptions.java:77)
	at com.twitter.intellij.pants.model.PantsOptions.lambda$getPantsOptions$0(PantsOptions.java:47)
	at java.util.Optional.map(Optional.java:215)
	at com.twitter.intellij.pants.model.PantsOptions.getPantsOptions(PantsOptions.java:47)
	at com.twitter.intellij.pants.file.FileChangeTracker.detectChangeType(FileChangeTracker.java:138)
	at com.twitter.intellij.pants.file.FileChangeTracker.markDirty(FileChangeTracker.java:110)
	at com.twitter.intellij.pants.file.FileChangeTracker.access$200(FileChangeTracker.java:46)
	at com.twitter.intellij.pants.file.FileChangeTracker$1.fileCreated(FileChangeTracker.java:315)
	at com.intellij.openapi.vfs.newvfs.VirtualFileFilteringListener.fileCreated(VirtualFileFilteringListener.java:69)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.util.EventDispatcher.dispatchVoidMethod(EventDispatcher.java:130)
	at com.intellij.util.EventDispatcher.access$000(EventDispatcher.java:24)
	at com.intellij.util.EventDispatcher$1.invoke(EventDispatcher.java:88)
	at com.sun.proxy.$Proxy24.fileCreated(Unknown Source)
	at com.intellij.openapi.vfs.impl.BulkVirtualFileListenerAdapter.fireAfter(BulkVirtualFileListenerAdapter.java:66)
	at com.intellij.openapi.vfs.impl.BulkVirtualFileListenerAdapter.after(BulkVirtualFileListenerAdapter.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.util.messages.impl.MessageBusImpl.invokeListener(MessageBusImpl.java:513)
	at com.intellij.util.messages.impl.MessageBusConnectionImpl.deliverMessage(MessageBusConnectionImpl.java:142)
	at com.intellij.util.messages.impl.MessageBusImpl.doPumpMessages(MessageBusImpl.java:438)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpWaitingBuses(MessageBusImpl.java:398)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpMessages(MessageBusImpl.java:388)
	at com.intellij.util.messages.impl.MessageBusImpl.sendMessage(MessageBusImpl.java:370)
	at com.intellij.util.messages.impl.MessageBusImpl.lambda$createTopicHandler$1(MessageBusImpl.java:241)
	at com.sun.proxy.$Proxy104.validityChanged(Unknown Source)
	at com.intellij.openapi.vfs.impl.VirtualFilePointerManagerImpl.after(VirtualFilePointerManagerImpl.java:658)
	at com.intellij.openapi.vfs.impl.VirtualFilePointerManagerImpl.access$600(VirtualFilePointerManagerImpl.java:50)
	at com.intellij.openapi.vfs.impl.VirtualFilePointerManagerImpl$4.afterVfsChange(VirtualFilePointerManagerImpl.java:567)
	at com.intellij.openapi.vfs.newvfs.AsyncEventSupport.afterVfsChange(AsyncEventSupport.java:109)
	at com.intellij.openapi.vfs.newvfs.AsyncEventSupport.access$200(AsyncEventSupport.java:27)
	at com.intellij.openapi.vfs.newvfs.AsyncEventSupport$1.after(AsyncEventSupport.java:53)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.util.messages.impl.MessageBusImpl.invokeListener(MessageBusImpl.java:513)
	at com.intellij.util.messages.impl.MessageBusConnectionImpl.deliverMessage(MessageBusConnectionImpl.java:142)
	at com.intellij.util.messages.impl.MessageBusImpl.doPumpMessages(MessageBusImpl.java:438)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpWaitingBuses(MessageBusImpl.java:398)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpMessages(MessageBusImpl.java:388)
	at com.intellij.util.messages.impl.MessageBusImpl.sendMessage(MessageBusImpl.java:372)
	at com.intellij.util.messages.impl.MessageBusImpl.lambda$createTopicHandler$1(MessageBusImpl.java:241)
	at com.sun.proxy.$Proxy103.after(Unknown Source)
	at com.intellij.openapi.vfs.newvfs.persistent.PersistentFSImpl.processEvent(PersistentFSImpl.java:698)
	at com.intellij.openapi.vfs.newvfs.persistent.PersistentFSImpl.createChildFile(PersistentFSImpl.java:453)
	at com.intellij.openapi.vfs.newvfs.impl.VirtualFileSystemEntry.createChildData(VirtualFileSystemEntry.java:212)
	at com.intellij.configurationStore.StorageUtilKt.getOrCreateVirtualFile(storageUtil.kt:133)
	at com.intellij.configurationStore.FileBasedStorageKt.writeFile(FileBasedStorage.kt:264)
	at com.intellij.configurationStore.FileBasedStorage$FileSaveSession.saveLocally(FileBasedStorage.kt:113)
	at com.intellij.configurationStore.XmlElementStorage$XmlElementStorageSaveSession$SaveExecutor.save(XmlElementStorage.kt:164)
	at com.intellij.configurationStore.SaveExecutorKt.executeSave(SaveExecutor.kt:90)
	at com.intellij.configurationStore.SaveExecutorKt.saveSessions(SaveExecutor.kt:84)
	at com.intellij.configurationStore.ProjectSaveSessionProducerManager$saveWithAdditionalSaveSessions$$inlined$withEdtContext$1$lambda$1.compute(actions.kt:80)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$runWriteAction$11(ApplicationImpl.java:894)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteActionWithClass(ApplicationImpl.java:873)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:894)
	at com.intellij.configurationStore.ProjectSaveSessionProducerManager$saveWithAdditionalSaveSessions$$inlined$withEdtContext$1.invokeSuspend(ComponentStoreImpl.kt:655)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:238)
	at com.intellij.configurationStore.EdtPoolDispatcherManager.processQueue(EdtPoolDispatcher.kt:54)
	at com.intellij.configurationStore.EdtPoolDispatcherManager.access$processQueue(EdtPoolDispatcher.kt:18)
	at com.intellij.configurationStore.EdtPoolDispatcherManager$scheduleFlush$1.invoke(EdtPoolDispatcher.kt:32)
	at com.intellij.configurationStore.EdtPoolDispatcherManager$scheduleFlush$1.invoke(EdtPoolDispatcher.kt:18)
	at com.intellij.configurationStore.EdtPoolDispatcherKt$sam$java_lang_Runnable$0.run(EdtPoolDispatcher.kt)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:309)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun$$$capture(LaterInvocator.java:441)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun(LaterInvocator.java)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:424)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:407)
	at java.awt.event.InvocationEvent.dispatch$$$capture(InvocationEvent.java:311)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:762)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:732)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:908)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:781)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:424)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:687)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:423)
	at com.intellij.ide.IdeEventQueue.pumpEventsForHierarchy(IdeEventQueue.java:965)
	at com.intellij.openapi.progress.util.ProgressWindow.startBlocking(ProgressWindow.java:186)
	at com.intellij.openapi.progress.util.ProgressWindow.startBlocking(ProgressWindow.java:172)
	at com.intellij.openapi.application.impl.ApplicationImpl.runProcessWithProgressSynchronously(ApplicationImpl.java:455)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressSynchronously(CoreProgressManager.java:453)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.runProcessWithProgressSynchronously(ProgressManagerImpl.java:82)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runSynchronously(CoreProgressManager.java:318)
	at com.intellij.openapi.progress.impl.CoreProgressManager.run(CoreProgressManager.java:303)
	at com.intellij.configurationStore.SaveAndSyncHandlerImpl.saveSettingsUnderModalProgress(SaveAndSyncHandlerImpl.kt:212)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.closeProject(ProjectManagerImpl.java:740)
	at com.intellij.openapi.project.impl.ProjectManagerImpl.closeAndDispose(ProjectManagerImpl.java:781)
	at com.intellij.openapi.wm.impl.CloseProjectWindowHelper.closeProjectAndShowWelcomeFrameIfNoProjectOpened(CloseProjectWindowHelper.kt:44)
	at com.intellij.openapi.wm.impl.CloseProjectWindowHelper.windowClosing(CloseProjectWindowHelper.kt:32)
	at com.intellij.openapi.wm.impl.ProjectFrameHelper$2.windowClosing(ProjectFrameHelper.java:282)
	at java.awt.AWTEventMulticaster.windowClosing(AWTEventMulticaster.java:349)
	at java.awt.AWTEventMulticaster.windowClosing(AWTEventMulticaster.java:349)
	at java.awt.Window.processWindowEvent(Window.java:2054)
	at javax.swing.JFrame.processWindowEvent(JFrame.java:305)
	at java.awt.Window.processEvent(Window.java:2013)
	at java.awt.Component.dispatchEventImpl(Component.java:4897)
	at java.awt.Container.dispatchEventImpl(Container.java:2295)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4719)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:737)
	at java.awt.EventQueue$4.run(EventQueue.java:735)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:908)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:781)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:424)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:698)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:423)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

### PR DESCRIPTION
**Pull request by tpasternak**
FileChangeTracker methods are always called in Event Dispatch Thread.
Pants Plugin implementation sometimes needs to retrieve pants options
in order to decide, whether to mark project as dirty. Unfortunatly,
Pants Options are retrieved via an external process call, what is
forbidden in the Event Dispatch Thread.

In this commit, the bahvior of the listener is changed in that way,
the pants options are retrieved before the listener is initialized,
on a thread from thread pool, and they're passed to listener's
constructor.

Unfortunately, pants options may be read only when a project
contains any modules. Therefore, it we can't just read them at
`projectOpened` stage, because the project may be initially empty.
That's why we register a ModuleListener in case when the project is
empty.

Fixes #463

### CODE REVIEW

File: src/com/twitter/intellij/pants/file/FileChangeTracker.java [hunk 1]
```diff
@@ -298,36 +297,41 @@ public static void unregisterProject(@NotNull Project project) {
       });
   }
 
-  private static VirtualFileListener getNewListener() {
+  private static VirtualFileListener getNewListener(final PantsOptions pantsOptions) {
     return new VirtualFileListener() {
+
+      void markDirty(VirtualFile file) {
```
Comment (lukaszwawrzyk) [inline src/com/twitter/intellij/pants/file/FileChangeTracker.java]: Can be private

File: src/com/twitter/intellij/pants/service/project/metadata/PantsMetadataService.java [hunk 2]
```diff
@@ -11,16 +12,23 @@
...
 import com.twitter.intellij.pants.util.PantsUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import com.google.gson.Gson;
 
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
```
Comment (lukaszwawrzyk) [inline src/com/twitter/intellij/pants/service/project/metadata/PantsMetadataService.java]: All these new imports are suspicious as there are no further changes in this file

File: src/com/twitter/intellij/pants/file/FileChangeTracker.java [hunk 3]
```diff
@@ -104,10 +104,10 @@ public FileChangeTracker getInstance() {
     return instance;
   }
 
-  private static void markDirty(@NotNull VirtualFile file, @NotNull VirtualFileListener listener) {
+  private static void markDirty(@NotNull VirtualFile file,final PantsOptions pantsOptions, @NotNull VirtualFileListener listener) {
```
Comment (lukaszwawrzyk) [inline src/com/twitter/intellij/pants/file/FileChangeTracker.java]: You can run formatter on this file

File: src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java [hunk 4]
```diff
@@ -188,6 +193,46 @@ private void prepareGuiComponents() {
           }
         }
 
+
+        /**
+         * VSF tracker implementation requires PantsOptions object, that cannot be constructed in
+         * the Event Dispatch Thread. What is more, it can't be constructed if the project contain
```
Comment (lukaszwawrzyk) [inline src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java]: contain**s**

File: src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java [hunk 5]
```diff
@@ -188,6 +193,46 @@ private void prepareGuiComponents() {
...
+         * no modules, and cannot be registered twice. Hence, there are two cases when VFS tracker is
+         * needed:
+         *   1. When the Pants project is opened and it already contains modules - then we can register tracker immediately
+         *   2. When the Pants project is opened, but modules are not loaded yet - then we have to wait until a module is loaded
+         * @param project
```
Comment (lukaszwawrzyk) [inline src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java]: This is redundant

File: src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java [hunk 6]
```diff
@@ -188,6 +193,46 @@ private void prepareGuiComponents() {
...
+         */
+        private void registerVfsListener(Project project) {
+          if(ModuleManager.getInstance(project).getModules().length > 0) {
```
Comment (lukaszwawrzyk) [inline src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java]: Also run formatter on this file

File: src/com/twitter/intellij/pants/file/FileChangeTracker.java (hunk 1 above)
Comment (tpasternak) [reply src/com/twitter/intellij/pants/file/FileChangeTracker.java]: Done

File: src/com/twitter/intellij/pants/service/project/metadata/PantsMetadataService.java (hunk 2 above)
Comment (tpasternak) [reply src/com/twitter/intellij/pants/service/project/metadata/PantsMetadataService.java]: Done

File: src/com/twitter/intellij/pants/file/FileChangeTracker.java (hunk 3 above)
Comment (tpasternak) [reply src/com/twitter/intellij/pants/file/FileChangeTracker.java]: Done

File: src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java (hunk 4 above)
Comment (tpasternak) [reply src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java]: Done

File: src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java (hunk 5 above)
Comment (tpasternak) [reply src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java]: Done

File: src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java (hunk 6 above)
Comment (tpasternak) [reply src/com/twitter/intellij/pants/components/impl/PantsProjectComponentImpl.java]: Done

File: src/com/twitter/intellij/pants/service/project/metadata/PantsMetadataService.java [hunk 7]
```diff
@@ -3,6 +3,7 @@
 
 package com.twitter.intellij.pants.service.project.metadata;
 
+import com.intellij.openapi.application.ApplicationManager;
```
Comment (wisechengyi) [inline src/com/twitter/intellij/pants/service/project/metadata/PantsMetadataService.java]: no change in this file.

**Pull request merged**